File: PythonAlgorithmInterView/선형자료구조/유효한괄호.js
/**
 * @param {string} s
 * @return {boolean}
 * 20번 문제.
 */
 var isValid = function(s) {

    let answer = true
    let stack = [];

    let table = {
        ')':'(',
        '}':'{',
        ']':'['
    }

    let charList = s.split('');

    charList.forEach( char => {
        if( !table[char]){
            stack.push(char)
        } else if( table[char] != stack.pop()){
            answer = false
        }
    })

    if(answer && stack.length == 0){
        return true;
    }
    return false
};


let s = "()"
console.log(isValid(s));

File: PythonAlgorithmInterView/선형자료구조/유효한괄호.py
def isValid(self, s: str) -> bool:
    stack = []
    table = {
        ')': '(',
        '}': '{',
        ']': '['
    }

    for char in s :
        if char not in table :
            stack.append(char)
        elif not stack or table[char] != stack.pop():
            return False
    print(stack)

    return len(stack) == 0


if __name__ == '__main__':
    s = "()["
    print(isValid(None, s))